### PR TITLE
gateway: optimize http route generation for gateways

### PR DESCRIPTION
**Pull request by ramaraochavali**
If gateway have multiple servers, we compute gateway routes for every server which is redundant. This PR computes routes once per gateway.
It also caches virtualservices for gateway so that we do not compute them per every server.
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ X] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
Benchmark results before and after

```
name                                old time/op        new time/op        delta
RouteGeneration/gateways-12                479ms ±10%         490ms ±10%     ~     (p=1.000 n=5+5)
RouteGeneration/gateways-shared-12         2.34s ± 7%         1.69s ± 6%  -27.62%  (p=0.008 n=5+5)
RouteGeneration/empty-12                  5.91ms ±10%        5.93ms ±14%     ~     (p=1.000 n=5+5)
RouteGeneration/tls-12                    5.55ms ± 2%        5.87ms ± 9%     ~     (p=0.190 n=4+5)
RouteGeneration/telemetry-12              6.27ms ±18%        5.96ms ±11%     ~     (p=1.000 n=5+5)
RouteGeneration/virtualservice-12         17.5ms ±10%        15.3ms ±12%  -12.15%  (p=0.032 n=5+5)

name                                old kb/msg         new kb/msg         delta
RouteGeneration/gateways-12                2.73k ± 0%         2.73k ± 0%     ~     (all equal)
RouteGeneration/gateways-shared-12         12.6k ± 0%         12.3k ± 0%   -2.69%  (p=0.008 n=5+5)
RouteGeneration/empty-12                    32.0 ± 0%          32.0 ± 0%     ~     (all equal)
RouteGeneration/tls-12                      32.0 ± 0%          32.0 ± 0%     ~     (all equal)
RouteGeneration/telemetry-12                32.0 ± 0%          32.0 ± 0%     ~     (all equal)
RouteGeneration/virtualservice-12           80.9 ± 0%          80.9 ± 0%     ~     (all equal)

name                                old resources/msg  new resources/msg  delta
RouteGeneration/gateways-12                1.00k ± 0%         1.00k ± 0%     ~     (all equal)
RouteGeneration/gateways-shared-12          2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/empty-12                    2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/tls-12                      2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/telemetry-12                2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/virtualservice-12           1.00 ± 0%          1.00 ± 0%     ~     (all equal)

name                                old alloc/op       new alloc/op       delta
RouteGeneration/gateways-12               15.7MB ± 0%        15.9MB ± 0%   +1.28%  (p=0.008 n=5+5)
RouteGeneration/gateways-shared-12        67.8MB ± 0%        37.4MB ± 0%  -44.91%  (p=0.000 n=4+5)
RouteGeneration/empty-12                   568kB ± 0%         568kB ± 0%     ~     (p=0.516 n=5+5)
RouteGeneration/tls-12                     567kB ± 0%         567kB ± 0%     ~     (p=0.190 n=5+4)
RouteGeneration/telemetry-12               568kB ± 0%         568kB ± 0%     ~     (p=0.548 n=5+5)
RouteGeneration/virtualservice-12          866kB ± 0%         866kB ± 0%     ~     (p=0.937 n=5+5)

name                                old allocs/op      new allocs/op      delta
RouteGeneration/gateways-12                 205k ± 0%          205k ± 0%   +0.01%  (p=0.008 n=5+5)
RouteGeneration/gateways-shared-12          902k ± 0%          511k ± 0%  -43.41%  (p=0.000 n=4+5)
RouteGeneration/empty-12                   5.82k ± 1%         5.81k ± 1%     ~     (p=0.762 n=5+5)
RouteGeneration/tls-12                     5.79k ± 0%         5.79k ± 0%     ~     (p=0.810 n=5+4)
RouteGeneration/telemetry-12               5.81k ± 0%         5.81k ± 0%     ~     (p=1.000 n=5+5)
RouteGeneration/virtualservice-12          11.0k ± 0%         11.0k ± 0%     ~     (p=0.429 n=4+4)
```

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 1]
```diff
@@ -243,10 +244,16 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(node *model.Pr
...
+			var routes []*route.Route
+			var exists bool
+			var err error
+			if routes, exists = gatewayRoutes[gatewayName]; !exists {
+				routes, err = istio_route.BuildHTTPRoutesForVirtualService(node, push, virtualService, nameToServiceMap, port, map[string]bool{gatewayName: true})
+				if err != nil {
+					log.Debugf("%s omitting routes for virtual service %v/%v due to error: %v", node.ID, virtualService.Namespace, virtualService.Name, err)
+					continue
+				}
+				gatewayRoutes[gatewayName] = routes
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:256]: Isn't the route releated to `virtualService`? Can not imagine how the ci passed

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/gateway.go:256]: thats a good point. Let me check

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: Fixed. PTAL

**Comment by ramaraochavali** [comment]
```
name                                old time/op        new time/op        delta
RouteGeneration/gateways-12                568ms ±32%         466ms ± 7%     ~     (p=0.222 n=5+5)
RouteGeneration/gateways-shared-12         2.66s ±33%         2.20s ±10%     ~     (p=0.421 n=5+5)
RouteGeneration/empty-12                  6.79ms ±32%        5.75ms ±10%     ~     (p=0.421 n=5+5)
RouteGeneration/tls-12                    6.92ms ±33%        5.43ms ± 3%  -21.51%  (p=0.016 n=5+5)
RouteGeneration/telemetry-12              6.98ms ±28%        5.79ms ±14%     ~     (p=0.151 n=5+5)
RouteGeneration/virtualservice-12         19.0ms ±28%        16.1ms ±22%     ~     (p=0.151 n=5+5)
```

File: pilot/pkg/xds/testdata/benchmarks/gateways.yaml [hunk 2]
```diff
@@ -67,6 +67,78 @@ spec:
...
+    tls:
+      mode: SIMPLE
+      serverCertificate: /etc/certs/servercert.pem
+      privateKey: /etc/certs/privatekey.pem
+  - port:
+      number: 443
+      name: https-simple-6
+      protocol: HTTPS
+    hosts:
+    - random-a-{{$i}}.host.example
```
Comment (howardjohn) [inline pilot/pkg/xds/testdata/benchmarks/gateways.yaml:135]: isn't this config invalid? we have overlapping host names?

Comment (ramaraochavali) [reply pilot/pkg/xds/testdata/benchmarks/gateways.yaml:135]: hmm. you are right. Let me revert this gateways.yaml change.

**Pull request merged**